geopy/geocoders/geoclient.py

```diff
@@ -0,0 +1,156 @@
+            self,
+            app_id=None,
+            app_key=None,
+            domain='api.cityofnewyork.us/geoclient/v1',
+            timeout=DEFAULT_TIMEOUT,
+            proxies=None,
+            user_agent=None,
+    ):
+        """
+        Initialize Geoclient geocoder. Please note that 'scheme' parameter is
+        not supported: at present state, all Geoclient traffic use https.
+
+        :param string appd_id: The application ID.
+
+        :param string app_key: The application key. 
+
+        :param string domain: Currently it is 'api.cityofnewyork.us/geoclient/v1', can
+            be changed if you have an on-premise Geoclient instance.
+
+        :param dict proxies: If specified, routes this geocoder's requests
+            through the specified proxy. E.g., {"https": "192.0.2.0"}. For
+            more information, see documentation on
+            :class:`urllib2.ProxyHandler`.
+
+        """
+        super(Geoclient, self).__init__(
+            scheme="https", timeout=timeout, proxies=proxies, user_agent=user_agent
+        )
+
+        self.app_id = app_id
+        self.app_key = app_key
+        
+        if not app_id:
+            raise ConfigurationError(
+                'No app_id given, required for api access.  If you do not '
+                'have a Geoclient app_id, sign up here: '
+                'https://developer.cityofnewyork.us/api/geoclient-api'
+                )
+        
+        if not app_key:
+            raise ConfigurationError(
+                'No app_key given, required for api access.  If you do not '
+                'have a Geoclient app_key, sign up here: '
+                'https://developer.cityofnewyork.us/api/geoclient-api'
+                )
+        
+        self.domain = domain.strip('/')
+        
+        '''use SFS (Single-field search) endpoint'''
+        self.geocode_api = 'https://%s/search.json' % (self.domain)
+
+    def geocode(
+            self,
+            query,
+            exactly_one=True,
+            timeout=None,
+            addressdetails=False,
+        ):
+        """
+        Geocode a location query.
+
+        :param string query: The query string to be geocoded using SFS syntax;
+            this must be URL encoded. Documentation at:
+            https://api.cityofnewyork.us/geoclient/v1/doc#section-1.3
+
+        :param bool exactly_one: Return one result or a list of results, if
+            available.
+
+        :param int timeout: Time, in seconds, to wait for the geocoding service
+            to respond before raising a :class:`geopy.exc.GeocoderTimedOut`
+            exception. Set this only if you wish to override, on this call
+            only, the value set during the geocoder's initialization.
+
+        :param bool addressdetails: If you want in *Location.raw* to include
+            address details such as BIN#, community district, etc set it to True.
+            Data Dictionary documenting all available fields at:
```

Comment:
I think the Python API should explicitly enumerate the available options by its parameters. This is a pass-through `**kwargs`, which is not as maintainable I think.